Your name: Thom Hemenway

In [76]:
def makettable(s1,s2):
    table = {}
    
    # Create table
    for i in range(len(s1)):
        table[s1[i]] = s2[i]
    
    return table


In [77]:
def trans(ttable, s):
    t = ""
    
    # Get character from table, otherwise just give us the character from the string
    for c in s:
        t += ttable.get(c, c)
    
    return t

In [78]:
# function to test translation code
# return True if successful, False if any test fails
def testtrans():
    ttable = makettable('abc', 'xyz')
    revttable = makettable('xyz', 'abc')
    tests = "Now I know my abc's"
    answer = "Now I know my xyz's"
    if trans(ttable, tests) != answer: return False
    if trans(revttable, trans(ttable, tests)) != "Now I know mb abc's": return False
    if trans(ttable,'') != '': return False
    if trans(makettable('',''), "abc") != 'abc': return False
    if trans(makettable('',''), "") != '': return False
    return True
testtrans()

True

## Comment
* The algorithm is well designed.
* The coding style is elegant.

In [2]:
Score = {}
Score["warmup"] = 30

In [3]:
# <summary>
#   Sorting algorithm to sort keys with same value
# </summary>
# <Parameters>
#   <l> List of pairs to sort</l>
#   <rev> Reverse variable for sorted() function</rev>
# </Parameters>
def sort_same_freq(l, rev):
    returnList = []
    listDi = []
    
    for (key, value) in l:
        if len(listDi) > 0:
            # Sort if we reached a new number of occurances
            if value != listDi[0][1]:
                # attach to return list
                returnList.extend(sorted(listDi, reverse=rev))
                listDi = []
        # Add tuple
        listDi.append((key,value))
    # Add last sorted list
    returnList.extend(sorted(listDi, reverse=rev))
    return returnList

In [4]:
def histo(s):
    d = {}
    
    # Make histogram
    for c in s:
        # Don't need newline or return line
        if (c != '\n') and (c != '\r'):
            d[c] = d.get(c, 0) + 1
    
    # Return sorted table
    return sort_same_freq(sorted(d.items(), key=lambda item: (item[1], item[0])), True)

In [5]:
def testhisto():
    if histo('implemented') == []: return False
    if histo('') != []: return False
    if histo('aacbbabccdedfefgih') == []: return False
    return True
testhisto()

True

## Bravo
## Comment
* The algorithm is well designed and thoughtful.
* The coding style is elegant.
* Good to see the assisted function sort_same_freq().
* The unit test is thorough.

In [6]:
Score["histo"] = 30
Score["testhisto"] = 5

In [88]:
def digraphs(s):
    l = {}
    
    # Create pair and add to dictionary
    for i in range(len(s)-1):
        if (s[i] != '\n') and (s[i] != '\r'):
            f = '/'+s[i]+s[i+1]+'/'
            l[f] = l.get(f, 0) + 1
    
    # sort by same value; sort same freq in increasing order
    return sort_same_freq(sorted(l.items(), key=lambda item: (item[1], item[0]), reverse=True), False)

In [89]:
def testdigraphs():
    print(digraphs("Where did Will go? He has her by the hand..."))
    print()
    print(digraphs("hahahahahahahahaha"))
    print()
    print(digraphs('a'))
    print()
    print(digraphs('......'))
    print(digraphs('.......'))
testdigraphs()

[('/ h/', 3), ('/e /', 3), ('/he/', 3), ('/../', 2), ('/er/', 2), ('/ha/', 2), ('/ H/', 1), ('/ W/', 1), ('/ b/', 1), ('/ d/', 1), ('/ g/', 1), ('/ t/', 1), ('/? /', 1), ('/He/', 1), ('/Wh/', 1), ('/Wi/', 1), ('/an/', 1), ('/as/', 1), ('/by/', 1), ('/d /', 1), ('/d./', 1), ('/di/', 1), ('/go/', 1), ('/id/', 1), ('/il/', 1), ('/l /', 1), ('/ll/', 1), ('/nd/', 1), ('/o?/', 1), ('/r /', 1), ('/re/', 1), ('/s /', 1), ('/th/', 1), ('/y /', 1)]

[('/ha/', 9), ('/ah/', 8)]

[]

[('/../', 5)]
[('/../', 6)]


## Comment
* The algorithm is well designed and thoughtful.
* The coding style is elegant.
* The unit test is thorough though, it misleads the direction of what the unit test should be. It should compare the output of digraph() with some groundtruth given. If they are equal, return true. Otherwise, return false. Here just print some outputs but who will check if the correctness of them?

In [7]:
Score["digraph"] = 30
Score["testdigraph"] = 4

In [90]:
crypt = """ZYX WXZVUTSRZXQ POZYNM SVLXK VM WRVM JNWXKZK, INNQSRMQK, RMQ MXRWHO
GWRKKSRMQK. VZ VK RSKN RKKNUVRZXQ IVZY WVLXWK RMQ VK JNTMQ VM RWXRK
IVZY MXRWHO KZWXRFK RMQ SREXK. RM XDUXSSXMZ KIVFFXW, VZ YRK XLXM HXXM
WXPNWZXQ JRW NTZ RZ KXR RMQ YRK UNMKXCTXMZSO UNSNMVBXQ FRMO KFRSS
VKSRMQK IVZYVM VZK WRMGX. QTWVMG ZYX XRWSO OXRWK NJ ZYX ZIXMZVXZY
UXMZTWO VZ VK KRVQ ZN YRLX HXXM UNFFNM XLXM VM HTKO PRWZK NJ HRMGENE,
KNFXZVFXK XRZVMG QNFXKZVU RMVFRSK.

ZYXVW MRZTWRS QVXZ VMUSTQXK FRFFRSK RMQ NUURKVNMRSSO HVWQK. KFRSS
KPXUVFXMK TP ZN 3–4 F (9.8–13.1 JZ) SNMG XRZ FRVMSO WNQXMZK KTUY RK
WRZK, IYXWXRK SRWGXW VMQVLVQTRSK KIVZUY ZN PWXO KTUY RK LVLXWWVQRX
(X.G. UVLXZK RMQ HVMZTWNMGK), RMQ XLXM PWVFRZXK RMQ PVGK. MXRW YTFRM
YRHVZRZVNM, ZYXO RWX EMNIM ZN KMRZUY KZWRO UYVUEXMK, URZK, RMQ QNGK NM
NUURKVNM. RFNMG ZYX SRWGXKZ, JTSSO QNUTFXMZXQ PWXO VZXFK ZN YRLX HXXM
ZREXM RWX R YRSJ-KZRWLXQ KTM HXRW NJ 23 EVSNGWRFK ZYRZ IRK XRZXM HO R
6.95 F (22.8 JZ) KPXUVFXM RMQ ZNNE KNFX ZXM IXXEK ZN QVGXKZ RK IXSS RK
PVGK NJ FNWX ZYRM 60 EG (132 SH). RK R WTSX NJ ZYTFH, ZYXKX KMREXK
KXXF RHSX ZN KIRSSNI PWXO TP ZN NMX-CTRWZXW ZYXVW NIM SXMGZY RMQ TP ZN
ZYXVW NIM IXVGYZ. RK IVZY RSS POZYNMK, ZYXO RWX PWVFRWVSO RFHTKY
YTMZXWK, TKTRSSO IRVZVMG TMZVS PWXO IRMQXWK IVZYVM KZWVEX WRMGX HXJNWX
KXVBVMG VZ VM ZYXVW UNVSK RMQ EVSSVMG LVR UNMKZWVUZVNM. YNIXLXW, ZYXWX
VK RZ SXRKZ NMX QNUTFXMZXQ URKX NJ R JNWRGVMG POZYNM XMZXWVMG R JNWXKZ
YTZ RMQ ZREVMG R UYVSQ. (XM.IVEVPXQVR.NWG/IVEV/POZYNM_WXZVUTSRZTK)"""

# Get rid of newline char
crypt = crypt.replace('\n', ' ')

# Initializing ttable
keytable = makettable('ZY', 'th')

# Added characters gradually
keytable["K"] = 's'
keytable["T"] = 'u'
keytable["M"] = 'n'
keytable["R"] = 'a'
keytable['X'] = 'e'
keytable["Q"] = 'd'
keytable["E"] = 'k'
keytable["S"] = 'l'
keytable["V"] = 'i'
keytable["P"] = 'p'
keytable['N'] = 'o'
keytable["U"] = 'c'
keytable["I"] = 'w'
keytable["G"] = 'g'
keytable['W'] = 'r'
keytable["J"] = 'f'
keytable["O"] = 'y'
keytable["H"] = 'b'
keytable["L"] = 'v'
keytable["F"] = 'm'
keytable["D"] = 'x'
keytable["C"] = 'q'
keytable["B"] = 'z'

# Print table
print("ENCRYPTION KEY")
for key, value in iter(keytable.items()):
    print("{0} -> {1}".format(key, value))

trans(keytable, crypt)

ENCRYPTION KEY
Z -> t
W -> r
L -> v
B -> z
M -> n
V -> i
E -> k
S -> l
F -> m
Q -> d
P -> p
C -> q
D -> x
Y -> h
R -> a
X -> e
H -> b
G -> g
T -> u
U -> c
O -> y
J -> f
K -> s
I -> w
N -> o


'the reticulated python lives in rain forests, woodlands, and nearby grasslands. it is also associated with rivers and is found in areas with nearby streams and lakes. an excellent swimmer, it has even been reported far out at sea and has consequently colonized many small islands within its range. during the early years of the twentieth century it is said to have been common even in busy parts of bangkok, sometimes eating domestic animals.  their natural diet includes mammals and occasionally birds. small specimens up to 3–4 m (9.8–13.1 ft) long eat mainly rodents such as rats, whereas larger individuals switch to prey such as viverridae (e.g. civets and binturongs), and even primates and pigs. near human habitation, they are known to snatch stray chickens, cats, and dogs on occasion. among the largest, fully documented prey items to have been taken are a half-starved sun bear of 23 kilograms that was eaten by a 6.95 m (22.8 ft) specimen and took some ten weeks to digest as well as pig

## Awesome
* Though you don't follow the instruction to demo the middle steps but just show the final answer. I give you FIVE!

In [10]:
Score["extra"] = 5

In [11]:
FinalScore = sum(Score.values())
print(FinalScore)

104


# Summary
## Good Job!
## FinalScore = 104
* The algorithms are well designed and thoughtful.
* The coding style is elegant.
* The testing is very thorough.
* The quality reaches the criteria of software engineering.